### DISCUSSION TITLE
.Net Core 7 backend developed but , it is not Working SSR Fetch in NextJS

### DISCUSSION THREAD

**Discussion by kubilaybzk** [vercel/next.js#53676] [Show and tell] [answered]
I have developed an ASP.NET Core 6 Web API. Now, I want to use this API in the frontend with NextJS. While server-side rendering, I cannot access the data and encounter errors, but I can access the data during client-side rendering. How can I solve this?

I used macbook therefore I use docker for my SQL Server:

`optionsBuilder.UseSqlServer("Data Source=192.168.16.1;Initial Catalog=EFCoreFirst;User ID=SA;Password=*************;TrustServerCertificate=True");`

This is my BackEnd's program.cs gist.
https://gist.github.com/kubilaybzk/a365fc53458f5438b311f74c9fd77415#file-program-cs

This is my SSR code:
![image](https://github.com/vercel/next.js/assets/48950686/ca569c73-7d9a-4d54-a0b4-896f864d2704)
This is my CSR code:
![image](https://github.com/vercel/next.js/assets/48950686/24872bf7-e7e2-433c-9c81-2bb91c31bd03)
This is my SSR result:
![image](https://github.com/vercel/next.js/assets/48950686/4a22bd8e-32b6-47f3-9fd6-a037819c048d)
This is console error:
![image](https://github.com/vercel/next.js/assets/48950686/3c955c6b-9760-4e60-be4d-9aa32e9b921c)
This is my terminal error .
![image](https://github.com/vercel/next.js/assets/48950686/bd1dc7ec-4146-4076-8e25-16c0ccfe1cd4)

I'm really researching and trying many techniques but haven't found a solution yet. I'm new to backend development with .Net, I was told to try methods like {mode:'no-cors'} in my research on the internet, I tried them but unfortunately I couldn't get a result.

My node version is : v20.5.0 Also my nextjs version is : "next": "13.4.12",

**Comment by masterkain** [comment]
`localhost` -- when running in a browser is one thing, when run in a docker environment is another

**Reply by kubilaybzk** [reply]
Okay but only sql is running on docker .Because as far as I know, I can't run mssql on mac in any other way, other than that, all my projects are running locally.

**Reply by masterkain** [reply]
but you need to expose the port from the running docker instance, try using compose maybe, it's a good idea anyway

**Comment by icyJoseph** [comment]
I guess, the .NET API is in the "same" network as your browser, but the Next.js app is not? How's the Next.js app running? Or rather in which environment is it started, and how?

**Reply by kubilaybzk** [reply]
Docker ,backend and Nextjs  work on the same computer  therefore same network.

**Reply by icyJoseph** [reply]
Oh, I had not read the other answer, only the DB is inside the docker instance. Alright.

Then I'd double check that localhost is resolvable, like try with `0.0.0.0:7039`, or `127.0.0.1:7039`, instead of localhost, inside of GSSP. This is most likely a networking issue, and we have to just double check every assumption.

**Reply by kubilaybzk** [reply]
<img width="704" alt="image" src="https://github.com/vercel/next.js/assets/48950686/7251a29c-4543-42c3-91d9-c8e0aadb0f35">

I have to try like this right?

`  const res = await fetch("https://127.0.0.1:7039/api/Products/GetAll",{mode:'no-cors'});
 `
 
<img width="694" alt="image" src="https://github.com/vercel/next.js/assets/48950686/1c67279f-2c86-447b-8d98-8408f5686589">

`  const res = await fetch("https://0.0.0.0:7039/api/Products/GetAll",{mode:'no-cors'});`


I have use this codes with my server component like this but again it show me same error.

I checked the results in NextJS in these four ways but the problem persists.
http://192.168.1.165:3000/Sunucu
http://0.0.0.0:3000/Sunucu
http://localhost:3000/Sunucu
http://127.0.0.1:3000/Sunucu

**Reply by icyJoseph** [reply]
Alright, we could put that aside for now.  

Do you see the requests arriving to your .NET backend? I wonder if there's some User Agent check failing, or any other kind of assertion done within it?

**Reply by kubilaybzk** [reply]
No, unfortunately I can't see any reaction.

**Reply by icyJoseph** [reply]
And do you see any more info on the response you get back? Any status code or text, any error raised by fetch? Maybe try catch the await of fetch and see what you get.

We should make sure that the server side fetch can reach that endpoint.

What if you make a script: test-fetch.js and try to fetch from that .NET backend, just copy the code into a function and run it. Log the results. 

then execute the script by running `node test-fetch.js` does that work?

**Reply by kubilaybzk** [reply]
![image](https://github.com/vercel/next.js/assets/48950686/355df175-1651-4f8d-b9aa-6babe49aa889)
Actually, I did  forget to add this, I get such an error when doing ssr in the browser tool.

**Reply by icyJoseph** [reply]
Did you try reaching the backend with a node script though? I wanna discard Next.js and possibly show that the issue is server to server communication.

**Reply by kubilaybzk** [reply]
I believe you're asking whether you need to perform this task from the "api" folder inside the "pages" directory, right?

**Reply by icyJoseph** [reply]
Not at all.

I'm asking if you can write a script, which - fetches from your .NET backend. 

Like forget about next.js for a moment. And just create a JavaScript file that does fetch from your backend. Run your backend and - run the script. Since you have node 20, fetch is native so it should just work. If it doesn't, then we can suspect that it is not a next.js issue.

**Reply by kubilaybzk** [reply]
Yes, I tried a method as you mentioned, in addition I created an html file containing html tags and javascript codes.
I then ran the project using LiveServer using VSCode.

`http://127.0.0.1:5500/index.html`

<img width="840" alt="image" src="https://github.com/vercel/next.js/assets/48950686/393c5984-e500-469d-a3fc-4fb27b830538">


I haven't had any problems here.
Additionally I wanted to run without server

`file:///Users/........./Desktop/index.html`
<img width="831" alt="image" src="https://github.com/vercel/next.js/assets/48950686/8e31237e-06f6-42a2-8bc3-1859bf8c0789">

I did not encounter any errors here.





    <table id="productTable" border="1">
        <thead>
            <tr>
                <th>ID</th>
                <th>Ürün Adı</th>
                <th>Stok</th>
                <th>Fiyat</th>
                <th>Oluşturma Zamanı</th>
            </tr>
        </thead>
        <tbody id="tableBody">
            
        </tbody>
    </table>

    <script>
        
        fetch("https://localhost:7039/api/Products/GetAll")
            .then(response => response.json())
            .then(products => {
                var tableBody = document.getElementById("tableBody");
                products.forEach(product => {
                    var row = document.createElement("tr");
                    row.innerHTML = `
                        <td>${product.id}</td>
                        <td>${product.name}</td>
                        <td>${product.stock}</td>
                        <td>${product.price}</td>
                        <td>${product.createTime}</td>
                    `;
                    tableBody.appendChild(row);
                });
            })
            .catch(error => {
                console.error("Error :", error);
            });
    </script>


`
I then created a JavaScript file as you mentioned.

First of all, I wanted to use axios.
`npm init -y`
Here I added the package.json and package-lock.js files required for my project and installed for axios.
I got  an Error : 
`Error fetching data: AxiosError: self-signed certificate` 

I add this configuration  top of the my code
`process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';`

<img width="501" alt="image" src="https://github.com/vercel/next.js/assets/48950686/6e7aac20-d11c-4ba9-bada-62edead9c461">

**Reply by kubilaybzk** [reply]
Now ı thing there is something about my SSL Configurations. Because if there is not an problem ı don't need to write this
`process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';` 
 command in my codes .

**Reply by icyJoseph** [reply]
The browser thing was a bit unnecessary. As said, you should focus on doing this from Node. 

Also you didn't need axios, since Node 18.xx native fetch is supported, so not even npm init was needed. 

How did you create your .NET HTTPS setup? Does it also work if you just reach it with `http://localhost:7039` etc etc?

Does using something like Node 18.16 help at all?

**Comment by airtonix** [comment]
the detail about MSQL and docker is irrelevant. remove that information or strike it out... it'll just confuse the situation.

1. you need to show us how you start your .net api server. 
2. your code shows you using `https://` but your testing urls are `http://`. ❗  are you running your server with a valid ssl cert? if not then just run the server without ssl.
3. you also need to prove that the api server works: 
    1. start the server
    2. in another terminal show the output of: 
        a. `curl http://localhost:7039` 
        a. `curl http://localhost:7039/api` 
        a. `curl http://localhost:7039/api/Products/GetAll` 
4. the image [you posted here](https://github.com/vercel/next.js/discussions/53676#discussioncomment-6662400) isn't really helpful. [HAR files are far more useful](https://support.google.com/admanager/answer/10358597?hl=en). (They're just JSON files, so once you've exported it from chrome, sift and clean it of sensitive info)

**Reply by kubilaybzk** [reply]
1-) I am running my application using VSCode for Mac as usual.
2-) I am conducting my tests both via https:// and http:// protocols.
3-)
![image](https://github.com/vercel/next.js/assets/48950686/a0708f16-5f12-48cb-ab34-e970bd97a073)
4-)This is my HAR file. https://drive.google.com/file/d/1cBdWokFUtK3tUFtALczxgmsdNHpvllff/view?usp=sharing

**Reply by airtonix** [reply]
you'll probably need to setup some another service on another port for non ssl that redirects to the ssl port. 


Once you do that then your attmepts to fetch from `http://...` will show: 
- a redirect response when fetched with `curl`
- a redirect and a follow up request to https in the browser.

**Reply by airtonix** [reply]
also for completenes: 

show us what the output of these are: 

```
curl http://localhost:7039
curl http://localhost:7039/api/
curl http://localhost:7039/api/Products
curl http://localhost:7039/api/Products/GetAll

curl https://localhost:7039
curl https://localhost:7039/api/
curl https://localhost:7039/api/Products
curl https://localhost:7039/api/Products/GetAll
```

**Reply by kubilaybzk** [reply, marked as answer]
I change my launch setting file , i have to remove all ssl configuration for it 

`{
  "$schema": "https://json.schemastore.org/launchsettings.json",
  "iisSettings": {
    "windowsAuthentication": false,
    "anonymousAuthentication": true,
    "iisExpress": {
      "applicationUrl": "http://localhost:13187",
      "sslPort": 0
    }
  },
  "profiles": {
    "http": {
      "commandName": "Project",
      "launchBrowser": true,
      "launchUrl": "swagger",
      "applicationUrl": "http://localhost:5031",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      },
      "dotnetRunMessages": true
    },
    "https": {
      "commandName": "Project",
      "launchBrowser": true,
      "launchUrl": "swagger",
      "applicationUrl": "http://localhost:7039;http://localhost:5031",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      },
      "dotnetRunMessages": true
    },
    "IIS Express": {
      "commandName": "IISExpress",
      "launchBrowser": true,
      "launchUrl": "swagger",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}`   

 Thanks for @sabriozay .

**Reply by kubilaybzk** [reply]
Here is Changes 
![image](https://github.com/vercel/next.js/assets/48950686/f666d274-3f37-4da1-bada-2c2caac465d5)
